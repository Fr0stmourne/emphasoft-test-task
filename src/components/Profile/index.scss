@import '../../scss/variables.scss';

.profile {
  display: grid;
  grid-template-columns: min-content auto;
  grid-template-rows: min-content auto;
  justify-items: center;
  align-items: center;
  column-gap: 20px;
  margin-bottom: 20px;
  font-family: $primary-font-family;

  &__photo {
    width: 150px;
    height: 150px;
    object-fit: cover;
    border-radius: 50%;
    grid-row: 1 / 3;

    @media (max-width: 500px) {
      width: 120px;
      height: 120px;
    }
  }

  &__name,
  &__logout {
    justify-self: center;
    text-align: center;
    height: fit-content;
    color: $link-color;
    text-decoration: underline;
    transition: 0.1s;

    &:visited,
    &:focus {
      color: $visited-link-color;
    }
  }

  &__logout {
    height: fit-content;
    padding: 0;
    border: 0;
    background-color: transparent;
    cursor: pointer;
    font-family: inherit;
  }
}