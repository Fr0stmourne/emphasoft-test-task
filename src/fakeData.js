const fakeFriends = [{
  "id": 1373072,
  "first_name": "Alex",
  "last_name": "Kouprin",
  "is_closed": true,
  "can_access_closed": false,
  "domain": "kouprin",
  "photo_100": "https://sun9-49.userapi.com/c855536/v855536573/24b6e1/Vx9ANB_8sos.jpg",
  "online": 0,
  "track_code": "9d0b2557-1EAZPZyjDnwU2i6oi5_r39uUCEOGXqb8ez4wBc6fGKWOlNVkyGPOfIGXxtMl-bfBm1QSGs"
  }, {
  "id": 5485951,
  "first_name": "Константин",
  "last_name": "Константинопольский",
  "is_closed": true,
  "can_access_closed": false,
  "domain": "yabalaban",
  "photo_100": "https://sun9-49.userapi.com/c855536/v855536573/24b6e1/Vx9ANB_8sos.jpg",
  "online": 1,
  "track_code": "db384e8d2lcIhKzG2_lZOUsppES_CpDQ8pvnHWhP5SXpQD0KLB63PFvlnsHW-Flmf4tK_SZ66dPy8oI"
  }, {
  "id": 84317567,
  "first_name": "Alexander",
  "last_name": "Bykov",
  "is_closed": false,
  "can_access_closed": true,
  "domain": "inited",
  "city": {
  "id": 1,
  "title": "Москва"
  },
  "photo_100": "https://sun9-49.userapi.com/c855536/v855536573/24b6e1/Vx9ANB_8sos.jpg",
  "online": 0,
  "track_code": "da23bda15d5C5fokKLhoYfKgL_EbEOgpwxqLYQutcGoDBjuUEzmItRyGkX59vG5hxQXwI45ziCTUAe4P"
  }]

const fakeProfile = {
  "id": 210700286,
  "first_name": "Константин",
  "last_name": "Константинопольский",
  "is_closed": false,
  "can_access_closed": true,
  "city": {
  "id": 5331,
  "title": "Los Angeles"
  },
  "photo_400_orig": 'https://sun9-49.userapi.com/c855536/v855536573/24b6e1/Vx9ANB_8sos.jpg',
  "verified": 1
};

export {
  fakeFriends, fakeProfile
}