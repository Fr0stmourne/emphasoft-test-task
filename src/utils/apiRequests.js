// import { fakeFriends, fakeProfile } from './fakeData';
import {
  proxy,
  redirectCallbackUrl,
  clientSecret,
  clientID,
} from './constants';

// function delay(ms) {
//   return new Promise((resolve, reject) => {
//     setTimeout(resolve, ms);
//   });
// }

export async function fetchFriends(token) {
  // await delay(2000);
  // return fakeFriends;

  try {
    const friendApi = `${proxy}https://api.vk.com/method/friends.get?count=5&order=random&fields=photo_100&v=5.122&&access_token=${token}`;
    const friendsResp = await fetch(friendApi);
    const data = await friendsResp.json();
    return data.response.items;
  } catch (e) {
    console.error(e);
  }
}

export async function fetchProfile(token) {
  // await delay(2000);
  // return fakeProfile;

  try {
    const userApi = `${proxy}https://api.vk.com/method/users.get?fields=photo_400_orig&v=5.122&&access_token=${token}`;
    const resp = await fetch(userApi);
    const data = await resp.json();
    return data.response[0];
  } catch (e) {
    console.error(e);
  }
}

export async function fetchToken(code) {
  try {
    const result = await fetch(
      `${proxy}https://oauth.vk.com/access_token?client_id=${clientID}&client_secret=${clientSecret}&redirect_uri=${redirectCallbackUrl}&code=${code}`
    );
    const json = await result.json();
    return json['access_token'];
  } catch (e) {
    console.error(e);
  }
}
