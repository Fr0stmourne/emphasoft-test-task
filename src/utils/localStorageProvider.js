import { tokenKey } from './constants';

export default class LocalStorageProvider {
  static setToken(token) {
    try {
      localStorage.setItem(tokenKey, token);
    } catch (e) {
      console.error(e);
    }
  }

  static getToken() {
    try {
      return localStorage.getItem(tokenKey);
    } catch (e) {
      console.error(e);
    }
  }

  static removeToken() {
    try {
      localStorage.removeItem(tokenKey);
    } catch (e) {
      console.error(e);
    }
  }
}
