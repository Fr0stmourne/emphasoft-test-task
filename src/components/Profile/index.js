import React from 'react';
import LocalStorageProvider from '../../utils/localStorageProvider';
import PropTypes from 'prop-types';
import './index.scss';
import { fakePhoto } from '../../utils/fakeData';

const defaultResource = {
  read() {
    return {
      first_name: 'Имя',
      last_name: 'Фамилия',
      photo_400_orig: fakePhoto,
      id: 0,
    };
  },
};

const logoutUser = () => {
  LocalStorageProvider.removeToken();
  document.location.reload();
};

export default function Profile({
  buttonText = 'Выйти',
  profile: profileResource = defaultResource,
}) {
  const {
    first_name: firstName,
    last_name: lastName,
    photo_400_orig: photoLink,
    id,
  } = profileResource.read();
  return (
    <div className="profile">
      <img
        className="profile__photo"
        src={photoLink}
        alt={`Аватар пользователя ${firstName} ${lastName}`}
      ></img>
      <a
        className="profile__name"
        href={`https://vk.com/id${id}`}
        target="_blank"
        rel="noopener noreferrer"
      >{`${firstName} ${lastName}`}</a>
      <button className="profile__logout" onClick={logoutUser}>
        {buttonText}
      </button>
    </div>
  );
}

Profile.propTypes = {
  buttonText: PropTypes.string,
  profileResource: PropTypes.object,
};
