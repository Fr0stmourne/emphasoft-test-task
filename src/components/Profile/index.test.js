import React from 'react';
import renderer from 'react-test-renderer';
import Profile from '.';
import { fakeProfile } from '../../utils/fakeData';

describe('Profile component', () => {
  test('should render the component correctly when no props are provided', () => {
    const tree = renderer.create(<Profile />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('should render the component with custom props correctly', () => {
    const resource = {
      read() {
        return fakeProfile;
      },
    };
    const tree = renderer
      .create(<Profile profileResource={resource} buttonText="Log out" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
