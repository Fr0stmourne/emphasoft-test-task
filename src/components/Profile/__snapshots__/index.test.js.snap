// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Profile component should render the component correctly when no props are provided 1`] = `
<div
  className="profile"
>
  <img
    alt="Аватар пользователя Имя Фамилия"
    className="profile__photo"
    src="https://sun9-49.userapi.com/c855536/v855536573/24b6e1/Vx9ANB_8sos.jpg"
  />
  <a
    className="profile__name"
    href="https://vk.com/id0"
    rel="noopener noreferrer"
    target="_blank"
  >
    Имя Фамилия
  </a>
  <button
    className="profile__logout"
    onClick={[Function]}
  >
    Выйти
  </button>
</div>
`;

exports[`Profile component should render the component with custom props correctly 1`] = `
<div
  className="profile"
>
  <img
    alt="Аватар пользователя Имя Фамилия"
    className="profile__photo"
    src="https://sun9-49.userapi.com/c855536/v855536573/24b6e1/Vx9ANB_8sos.jpg"
  />
  <a
    className="profile__name"
    href="https://vk.com/id0"
    rel="noopener noreferrer"
    target="_blank"
  >
    Имя Фамилия
  </a>
  <button
    className="profile__logout"
    onClick={[Function]}
  >
    Log out
  </button>
</div>
`;
